tests/test_llmblock.py

```diff
@@ -357,7 +360,95 @@ def test_resolve_model_family(self, mock_load_config):
+        self.ctx.model_family = old_model_family
+
+    def test_resolve_client(self, mock_load_config):
+        mock_load_config.return_value = {
+            "system": "{{fruit}}",
+            "introduction": "introduction",
+            "principles": "principles",
+            "examples": "examples",
+            "generation": "generation",
+        }
+        # save state for future tests
+        old_client = self.ctx.client
+        old_clients = self.ctx.clients
+
+        # Uses the default client when none specified
+        block = LLMBlock(
+            ctx=self.ctx,
+            pipe=self.mock_pipe,
+            block_name="gen_knowledge",
+            config_path="",
+            output_cols=[],
+        )
+        assert block.client == self.ctx.client
+
+        # client on block is used when given
+        client2 = MagicMock()
+        self.ctx.clients = {
+            "client2": client2,
+        }
+        block = LLMBlock(
+            ctx=self.ctx,
+            pipe=self.mock_pipe,
+            block_name="gen_knowledge",
+            config_path="",
+            output_cols=[],
+            client="client2",
+        )
+        assert block.client == client2
+
+        # raises an error when a block tries to use a client that's
+        # not defined
+        with pytest.raises(BlockConfigParserError) as exc:
+            LLMBlock(
+                ctx=self.ctx,
+                pipe=self.mock_pipe,
+                block_name="gen_knowledge",
+                config_path="",
+                output_cols=[],
+                client="client3",
+            )
+        assert "no client of that name" in str(exc.value)
+
+        # even if that's the "default" client, if the user gave an
+        # explicit mapping of clients that did not include a default
+        with pytest.raises(BlockConfigParserError) as exc:
+            LLMBlock(
+                ctx=self.ctx,
+                pipe=self.mock_pipe,
+                block_name="gen_knowledge",
+                config_path="",
+                output_cols=[],
+                client="default",
+            )
+        assert "no client of that name" in str(exc.value)
+
+        # can find the default client if default explicitly specified
+        client_default = MagicMock()
+        self.ctx.clients = {"default": client_default}
+        block = LLMBlock(
+            ctx=self.ctx,
+            pipe=self.mock_pipe,
+            block_name="gen_knowledge",
+            config_path="",
+            output_cols=[],
+            client="default",
+        )
+        assert block.client == client_default
+        block = LLMBlock(
+            ctx=self.ctx,
+            pipe=self.mock_pipe,
+            block_name="gen_knowledge",
+            config_path="",
+            output_cols=[],
+        )
+        assert block.client == client_default
+
+        # restore state for future tests
+        self.ctx.clients = old_clients
```

Comment:
How come we are relying on the inner logic of a particular test to preserve the integrity of other tests in this suite? Maybe I'm misunderstanding something, but it seems like this violates the encapsulation property that we would want this test to have.